src/lib/propagation/DAG.js

```diff
@@ -0,0 +1,63 @@
+import * as NodeFns from './nodeFns'
+import {SimulationNode} from './node'
+import * as constants from './constants'
+
+import * as _collections from 'gEngine/collections'
+
+export class SimulationDAG {
+  constructor(nodes) {
+    if (!!_.get(window, 'recorder')) { window.recorder.recordSimulationDAGConstructionStart(this) }
+    let rest = nodes.map(NodeFns.extractInputs)
+
+    let byId = _.transform(rest, (map, node) => {map[node.id] = {node, lastAncestors: node.inputs, ancestors: node.inputs}}, {})
+
+    const missingInputsNodes = _.remove(rest, NodeFns.hasMissingInputs(rest))
+    let graphErrorNodes = missingInputsNodes.map(NodeFns.withMissingInputError(nodes))
```

Comment:
above, I'd rename ``graphErrorNodes`` to ``missingInputNodesWithErrors`` or similar, if that's what they are.  Or, ``nodesWithErrors``